### PR TITLE
Kasli: new flash feature using core device management tool

### PR DESCRIPTION
**Pull request by SidaChen1999**
# ARTIQ Pull Request

## Description of Changes
Now the users of Kasli can flash the gateware/bootloader/firmware and reload the whole system using core device management tool. Commands are very similar to artiq_flash, besides adding the `flash` keyword after aritq_coremgmt. If no partition is specified, the default is `gateware bootloader firmware start`. 
Example: `artiq_coremgmt flash --srcbuild -d artiq_kasli -V <your_variant>`

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### All Pull Requests

- [x] Use correct spelling and grammar.

### Code Changes

- [x] Test your changes or have someone test them. Mention what was tested and how.

### Documentation Changes

- [ ] Check, test, and update the documentation in [doc/](../doc/). Build documentation (`cd doc/manual/; make html`) to ensure no errors.

### Git Logistics

- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,6 +8,9 @@ __pycache__/
 *.fbi
 *.pcap
 *.prof
+.vscode
+.direnv
+.envrc
```
Comment (sbourdeauducq) [inline .gitignore:13]: No, we don't put IDE-specific rules in gitignore.
(And a number of things should also be removed from this file e.g. *.pcap, *.prof,  .ipynb_checkpoints...)

File: artiq/coredevice/comm_mgmt.py [hunk 2]
```diff
@@ -24,11 +24,11 @@ class Request(Enum):
     StopProfiler = 10
     GetProfile = 11
 
-    Hotswap = 4
-    Reboot = 5
-
     DebugAllocator = 8
 
+    FlashWrite = 16
+    Relaod = 17
```
Comment (sbourdeauducq) [inline artiq/coredevice/comm_mgmt.py:30]: Fix typo

File: artiq/coredevice/comm_mgmt.py [hunk 3]
```diff
@@ -118,9 +119,11 @@ def _read_header(self):
         return ty
 
     def _read_expect(self, ty):
-        if self._read_header() != ty:
+        self._read_type = self._read_header()
+        if self._read_type != ty:
             raise IOError("Incorrect reply from device: {} (expected {})".
                           format(self._read_type, ty))
+        return True
```
Comment (sbourdeauducq) [inline artiq/coredevice/comm_mgmt.py:126]: These changes to ``_read_expect``are unnecessary, remove.

File: artiq/coredevice/comm_mgmt.py [hunk 4]
```diff
@@ -224,14 +227,27 @@ def get_profile(self):
...
-
     def debug_allocator(self):
         self._write_header(Request.DebugAllocator)
+    
+    def flash_write(self, partition, firmware):
+        self._write_header(Request.FlashWrite)
+        self._write_string(partition)
+        self._write_bytes(firmware)
+        if self._read_expect(Reply.RebootImminent):
+            print(partition, "write success")
```
Comment (sbourdeauducq) [inline artiq/coredevice/comm_mgmt.py:238]: Use logging like the rest of the code.

File: artiq/coredevice/comm_mgmt.py [hunk 5]
```diff
@@ -224,14 +227,27 @@ def get_profile(self):
...
+                print("Reload failed")
+                return False
+        except IOError:
+            return True
+
+    def awake(self):
+        self._write_header(Request.Awake)
+        if self._read_expect(Reply.Success):
+            return True
+        return False
```
Comment (sbourdeauducq) [inline artiq/coredevice/comm_mgmt.py:253]: EOL

File: artiq/firmware/libboard_misoc/flash.rs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &Error::AlreadyLocked =>
+                write!(f, "attempt at reentrant access"),
+            &Error::NoFlash =>
+                write!(f, "flash memory is not present"),
+            &Error::WrongPartition =>
+                write!(f, "Unknown partition"),
```
Comment (sbourdeauducq) [inline artiq/firmware/libboard_misoc/flash.rs:22]: Capitalize messages consistently

File: artiq/firmware/libboard_misoc/flash.rs [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    use csr;
+    use byteorder::{ByteOrder, BigEndian};
+    use flash::crc::crc32;
+
+    const SIZE: usize = spiflash::SECTOR_SIZE;
+    const ADDR_GATEWARE: usize = 0x0;
+    const ADDR_BOOTLOADER: usize = 0x400000;
+    const ADDR_FIRMWARE: usize = ::mem::FLASH_BOOT_ADDRESS;
+
+    mod lock {
```
Comment (sbourdeauducq) [inline artiq/firmware/libboard_misoc/flash.rs:49]: Unnecessary AFAICT and in any case would not belong here.

File: artiq/firmware/libboard_misoc/flash.rs [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            value = &value[size..];
+            if value.len() <= 0 {
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn reload () -> Result<(), Error> {
+
```
Comment (sbourdeauducq) [inline artiq/firmware/libboard_misoc/flash.rs:132]: Remove blank line.

File: artiq/firmware/libboard_misoc/flash.rs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        unsafe {
+            csr::icap::trigger_write(0);
```
Comment (sbourdeauducq) [inline artiq/firmware/libboard_misoc/flash.rs:134]: Use ``CSR`` instead of ``CSRStorage`` in MiSoC so only 1 write is necessary.

Since this module does reboot, names should be more explicit than "trigger icap".

File: artiq/firmware/libboard_misoc/flash.rs [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    pub fn write(key: &str, mut value: &[u8]) -> Result<(), Error> {
+        Err(Error::NoFlash)
+    }
+
+    pub fn reload () -> Result<(), Error> {
+        Err(Error::NoFlash)
+    }
+}
+
+pub use self::imp::*;
```
Comment (sbourdeauducq) [inline artiq/firmware/libboard_misoc/flash.rs:159]: EOL

File: artiq/firmware/libboard_misoc/lib.rs [hunk 11]
```diff
@@ -26,7 +27,7 @@ pub mod uart;
 #[cfg(has_spiflash)]
 pub mod spiflash;
 pub mod config;
-#[cfg(feature = "uart_console")]
+// #[cfg(feature = "uart_console")]
```
Comment (sbourdeauducq) [inline artiq/firmware/libboard_misoc/lib.rs:30]: Undo this change

File: artiq/frontend/artiq_coremgmt.py [hunk 12]
```diff
@@ -201,7 +204,92 @@ def main():
...
+                variant = variants[0]
+            else:
+                raise ValueError("more than one variant found for selected board, specify -V. "
+                    "Found variants: {}".format(" ".join(sorted(variants))))
+        variant_dir = ''
+        if needs_artifacts:
+            if args.srcbuild:
+                variant_dir = variant
+            else:
+                variant_dir = 'kasli-' + variant
```
Comment (sbourdeauducq) [inline artiq/frontend/artiq_coremgmt.py:235]: Do not duplicate this code from artiq_flash; share it.

File: artiq/frontend/artiq_coremgmt.py [hunk 13]
```diff
@@ -201,7 +204,92 @@ def main():
...
+                    bin_file.write(b"\x00"*8)
+                bit2bin(bit_file, bin_file)
+                if header:
+                    magic = 0x5352544d  # "SRTM", see sayma_rtm target
+                    length = bin_file.tell() - 8
+                    bin_file.seek(0)
+                    bin_file.write(magic.to_bytes(4, byteorder="big"))
+                    bin_file.write(length.to_bytes(4, byteorder="big"))
+            atexit.register(lambda: os.unlink(bin_filename))
+            return bin_filename
```
Comment (sbourdeauducq) [inline artiq/frontend/artiq_coremgmt.py:264]: ditto

File: artiq/frontend/artiq_coremgmt.py [hunk 14]
```diff
@@ -201,7 +204,92 @@ def main():
...
+                with open(bootloader_bin, "rb") as fi:
+                    file = fi.read()
+                    mgmt.flash_write(partition, file)
+            elif partition == "firmware":
+                firmware = "runtime"
+                firmware_fbi = artifact_path(variant_dir, "software", firmware, firmware + ".fbi")
+                with open(firmware_fbi, "rb") as fi:
+                    file = fi.read()
+                    mgmt.flash_write(partition, file)
+            elif partition == "start":
```
Comment (sbourdeauducq) [inline artiq/frontend/artiq_coremgmt.py:284]: partition?

File: artiq/frontend/artiq_coremgmt.py [hunk 15]
```diff
@@ -201,7 +204,92 @@ def main():
...
+                print('Reloading')
+                if mgmt.reload():
+                    mgmt.close()
+                    if(mgmt.awake()):
```
Comment (sbourdeauducq) [inline artiq/frontend/artiq_coremgmt.py:288]: Not needed, and AFAICT broken due to race condition.

File: artiq/gateware/targets/kasli.py [hunk 16]
```diff
@@ -131,6 +131,10 @@ def __init__(self, gateware_identifier_str=None, **kwargs):
                 self.platform.request("error_led")))
             self.csr_devices.append("error_led")
             self.submodules += SMAClkinForward(self.platform)
+        
+        if self.platform.hw_rev == "v2.0":
```
Comment (sbourdeauducq) [inline artiq/gateware/targets/kasli.py:135]: Why v2.0 only?

**Comment by sbourdeauducq** [comment]
Also test KC705 and Sayma.

File: artiq/frontend/artiq_coremgmt.py (hunk 12 above)
Comment (jbqubit) [reply artiq/frontend/artiq_coremgmt.py:235]: If the artiq_flash functionality is now part of artiq_coremgmt why not deprecate artiq_flash?

Comment (sbourdeauducq) [reply artiq/frontend/artiq_coremgmt.py:235]: Because artiq_coremgmt doesn't work if there isn't functional firmware already running.

File: artiq/frontend/artiq_coremgmt.py (hunk 14 above)
Comment (SidaChen1999) [reply artiq/frontend/artiq_coremgmt.py:284]: I think gateware/bootloader/firmware is like partitions in flash. I can change the name to `action` just like artiq_flash

File: artiq/firmware/libboard_misoc/flash.rs (hunk 7 above)
Comment (SidaChen1999) [reply artiq/firmware/libboard_misoc/flash.rs:49]: Do you mean the mod lock? It is from the config.rs, but I modified the functions so I cannot use that one.

Comment (sbourdeauducq) [reply artiq/firmware/libboard_misoc/flash.rs:49]: Ah, I see, it's because of the weird thread implementation we have.
I still don't see why it really has to be modified and it would be better to share it with config.rs.

Comment (SidaChen1999) [reply artiq/firmware/libboard_misoc/flash.rs:49]: In config.rs, the address is fixed to storage. But in my case, the address could be one of gateware/bootloader/firmware. So I added the address as a parameter. I can change this part in config.rs and share it.

File: artiq/frontend/artiq_coremgmt.py (hunk 15 above)
Comment (SidaChen1999) [reply artiq/frontend/artiq_coremgmt.py:288]: Should I remove `mgmt.awak()`? I think with it the user can get a better idea of the state.

Comment (sbourdeauducq) [reply artiq/frontend/artiq_coremgmt.py:288]: Yes, please remove it IMO.

File: artiq/firmware/libboard_misoc/flash.rs (hunk 9 above)
Comment (SidaChen1999) [reply artiq/firmware/libboard_misoc/flash.rs:134]: I'm not sure which signal I should use, `r` or `re`. But I tested that 1 write also works for `CSRStorage`

Comment (sbourdeauducq) [reply artiq/firmware/libboard_misoc/flash.rs:134]: > I'm not sure which signal I should use, r or re.

``re`` is pulsed when the CPU write the register, and ``r`` contains the value (and is only valid with ``re`` asserted). So use ``re``.

> But I tested that 1 write also works for CSRStorage

Maybe, but you should use the right thing.
